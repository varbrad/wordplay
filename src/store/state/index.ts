// Primitive types
export type Id = string | number

export interface Identifiable {
  id: Id
}

export enum TileType {
  STANDARD,
  WILDCARD,
}

export interface StandardTile extends Identifiable {
  type: 'standard'
  letter: string
  value: number
}

export interface WildcardTile extends Identifiable {
  type: 'wildcard'
  letter?: string
}

// State types
export type Tile = StandardTile | WildcardTile

export interface Rack {
  id: Id
  tiles: Tile[]
}

export interface StoreState {
  racks: Rack[]
}

export const createInitialState = (
  initialState?: Partial<StoreState>,
): StoreState => ({
  racks: [],
  ...initialState,
})
