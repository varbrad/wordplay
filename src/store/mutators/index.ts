import { SetState } from 'zustand'
import { StoreState } from '../state'

import simple from './simple'

export type StoreSetter = SetState<StoreState>

type RetT = <T extends FunctionConstructor>(a: T) => ReturnType<T>

export const createMutators = (set: StoreSetter) => ({
  set: (state: Partial<StoreState>, preserve = true): void =>
    set(preserve ? state : () => state),

  // Mutations
  simple: simple(set),
})

export type StoreMutations = ReturnType<typeof createMutators>
