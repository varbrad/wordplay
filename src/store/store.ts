import create from 'zustand'
import { createInitialState, StoreState } from './state'
import { createMutators, StoreMutations } from './mutators'

type Store = StoreState & StoreMutations

export const makeStore = (initialState?: Partial<StoreState>) =>
  create<Store>(set => ({
    // Make our mutators
    ...createMutators(set),
    ...createInitialState(initialState),
  }))
