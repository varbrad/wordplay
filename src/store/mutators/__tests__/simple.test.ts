import { makeStore } from '../../store'

describe('simple mutator', () => {
  it('should add a new tile to the rack', () => {
    const [, store] = makeStore()

    expect(store.getState().racks).toHaveLength(0)

    store.getState().simple()

    expect(store.getState().racks).toMatchSnapshot()
  })
})
