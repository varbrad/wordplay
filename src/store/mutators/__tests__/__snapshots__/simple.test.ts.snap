// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple mutator should add a new tile to the rack 1`] = `
Array [
  Object {
    "id": "a",
    "tiles": Array [],
  },
]
`;
